### ISSUE TITLE
Cloning `Tts` causes a double free

### ISSUE THREAD

**Issue by jquesada2016** [ndarilek/tts-rs#23]
Hello! Thanks again for this crate!

I noticed that cloning a `Tts` instance causes a double free error for me on Linux.

Given the following example:

```rust
use futures::task::{LocalSpawn, LocalSpawnExt, Spawn};
use std::cell::RefCell;
use std::ops::DerefMut;
use std::pin::Pin;
use std::rc::Rc;
use std::{
    future::Future,
    task::{Context, Poll},
};
use tts::{Backends, Tts};

/// Trait to implement async support for [`Tts`].
trait TtsExt {
    fn speak_async<'a>(&mut self, text: &'a str, interrupt: bool) -> SpeakAsync<'a>;
}

impl TtsExt for Tts {
    fn speak_async<'a>(&mut self, text: &'a str, interrupt: bool) -> SpeakAsync<'a> {
        SpeakAsync {
            tts: self.clone(),
            text,
            interrupt,
            first_call: Rc::new(RefCell::new(true)),
        }
    }
}

/// Struct which allows awaiting an utterance.
pub struct SpeakAsync<'a> {
    tts: Tts,
    text: &'a str,
    interrupt: bool,
    first_call: Rc<RefCell<bool>>,
}

impl<'a> Future for SpeakAsync<'a> {
    type Output = Result<(), tts::Error>;

    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        // If this is our first time being polled, register the callback
        if *self.first_call.borrow() {
            let waker = cx.waker().to_owned();
            let first_call = self.first_call.clone();

            self.tts
                .on_utterance_end(Some(Box::new(move |_| {
                    *first_call.borrow_mut() = false;

                    // We want to wake by ref to avoid consuming waker,
                    // or cloning it
                    waker.wake_by_ref()
                })))
                .expect("Failed to register `on_utterance_end` callback");

            // We need to move these items out of `self` to avoid
            // having a mutable borrow in the same scope as
            // immutable borrows
            let text = self.text;
            let interrupt = self.interrupt;

            match self.tts.speak(text, interrupt) {
                Ok(_) => Poll::Pending,
                Err(err) => Poll::Ready(Err(err)),
            }
        }
        // Otherwise, this means we have been woken up by an
        // utterence ended event, and are therefore done
        else {
            Poll::Ready(Ok(()))
        }
    }
}

pub fn speak(text: &str) {
    // Specify the TTS backend that will be used
    #[cfg(target_os = "android")]
    let backend = Backends::Android;
    #[cfg(any(target_os = "macos", target_os = "ios"))]
    let backend = Backends::AvFoundation;
    #[cfg(target_os = "linux")]
    let backend = Backends::SpeechDispatcher;
    #[cfg(all(windows, feature = "tolk"))]
    let backend = Backends::Tolk;
    #[cfg(target_arch = "wasm32")]
    let backend = Backends::Web;
    #[cfg(windows)]
    let backend = Backends::WinRt;

    // Create Tts instance
    let mut tts = Tts::new(backend).expect("Failed to initialize TTS");

    // Speak text
    futures::executor::block_on(tts.speak_async(text, false))
        .unwrap_or_else(|_| panic!(r#"Failed to speak "{}""#, text));

    println!(r#"Done speaking "{}""#, text);
}


fn main() {
    speak("Hello, this is a test");
}
```

results in a double free error and aborts.

**Comment by ndarilek** [comment]
Do you have an actual stacktrace for the crash?

Also, can you duplicate this with a simpler example? I.e. cloning tts in 
hello_world?

**Comment by jquesada2016** [comment]
Sure, let me go ahead and do that for you.

**Comment by jquesada2016** [comment]
[here](https://github.com/jquesada2016/tts-clone-unsoundness) is the repro repo.

This is the output I get, it doesn't seem to be a rust crash, so don't really know how to generate a stack trace.
```
    Finished dev [unoptimized + debuginfo] target(s) in 0.70s
     Running `target/debug/tts-clone-unsoundness`
free(): double free detected in tcache 2
Aborted
```

**Comment by ndarilek** [comment]
I'm unfortunately not able to reproduce this under Windows. However, [this branch](https://github.com/ndarilek/tts-rs/tree/soundness) stores the backend in an `Arc<RwLock<...>>` so hopefully the contained backend only gets dropped once. I've only made the switch on Linux and Windows for now, but if it works for you then I'll do it for the remaining platforms and push out a new release. Mind giving it a shot?

**Comment by jquesada2016** [comment]
I have tried the `soundness` branch, and it does not produce a double free error!

**Comment by jquesada2016** [comment]
I also forgot to mention, I am running Linux on WSL2 on Windows 11.

**Comment by ndarilek** [comment]
Awesome, thanks for confirming that it works. Unfortunately I don't have 
sound in WSL here so I can't run the Linux examples.

Mind running one more check? I just pushed a clone_drop example that 
clones and drops a `Tts` instance since I needed to work around `Drop` 
behavior and callback cleanup. Could you please pull and run `cargo run 
--example clone_drop` and ensure that works in WSL? If it does, I'll 
merge this branch and push 0.20.5.

**Comment by jquesada2016** [comment]
I just ran the example you mantioned and got the following output:

```
    Finished dev [unoptimized + debuginfo] target(s) in 25.51s
     Running `target/debug/examples/clone_drop`
No screen reader is available on this platform.
Are we speaking? false
Started speaking SpeechDispatcher(1)
Finished speaking SpeechDispatcher(1)
Started speaking SpeechDispatcher(2)
Finished speaking SpeechDispatcher(2)
Started speaking SpeechDispatcher(3)
Finished speaking SpeechDispatcher(3)
Started speaking SpeechDispatcher(4)
Finished speaking SpeechDispatcher(4)
Started speaking SpeechDispatcher(5)
Finished speaking SpeechDispatcher(5)
Started speaking SpeechDispatcher(6)
Finished speaking SpeechDispatcher(6)
Started speaking SpeechDispatcher(7)
Finished speaking SpeechDispatcher(7)
Started speaking SpeechDispatcher(8)
Finished speaking SpeechDispatcher(8)
Started speaking SpeechDispatcher(9)
Finished speaking SpeechDispatcher(9)
Started speaking SpeechDispatcher(10)
Finished speaking SpeechDispatcher(10)
Started speaking SpeechDispatcher(11)
Finished speaking SpeechDispatcher(11)
Started speaking SpeechDispatcher(12)
Finished speaking SpeechDispatcher(12)
```

At this point, the program hangs indefinitely and does not exit or continue. (haven't looked at the source for the example, however)

**Comment by ndarilek** [comment]
That is the correct behavior. Thanks for confirming that it works.

I also fixed this lower in the stack in speech-dispatcher-rs. New 
versions of both with this fix should go live later today.

**Comment by ndarilek** [comment]
This should be fixed in 0.21.0, wending its way through CI now. Thanks for the report, and for help testing!

Note that this version also includes explicit support for speech-dispatcher 0.10 and above. You aren't reporting build failures, which suggests you're still on 0.9. When you upgrade, you'll probably need to disable default features or upgrade speech-dispatcher.
